carabiner-to-py/CarabinerToPy.py

```diff
@@ -0,0 +1,80 @@
+import socket
+import edn_format
+import time
+
+class LinkListener():
+    """A simple python client to communicate with carabiner (a Abelton Link connector)
+    Carabiner server must be running to use. Requires edn_format [$pip install edn_format]"""
+    def __init__(self, tcp_ip = '127.0.0.1', tcp_port = 17000, buffer_size = 1024):
+        self.tcp_ip = tcp_ip
+        self.tcp_port = tcp_port
+        self.buffer_size = buffer_size
+        self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.s.connect((tcp_ip, tcp_port))
+
+    def decode_edn_msg(self, msg):
+        """Decodes a TCP message from Carabiner to python dictionary"""
+        msg = msg.decode()
+        striped_msg = msg[msg.index('{'):]
+        decoded_msg = edn_format.loads(striped_msg)
+
+        # Because the edf_format package does not return normal dam dicts (or string keywords). What dicts.
+        if type(decoded_msg) is edn_format.immutable_dict.ImmutableDict:
+            decoded_dict = {}
+            for key, value in decoded_msg.dict.items():
+                print()
+                decoded_dict[str(key).strip(':')] = value
+            decoded_msg = decoded_dict
+
+        return decoded_msg
+
+    def status(self):
+        """Wrapper for Status"""
+        self.s.send(b'status')
+        return_msg = self.s.recv(self.buffer_size)
+        return self.decode_edn_msg(return_msg)

```

Comment:
Thanks for adding those extra functions, I think this nicely completes the Carabiner package :)

Ah, so Caribiner pushing TCP messages when Link tempo, beats, etc change? I was under the impression that the only way to get information from the link session was to explicitly ask for it. In the former case, I agree a listener thread/callback system is a good idea!
I'm not that familiar with TCP, but do you think that all replies from carabiner should be read by a listener thread, even when these messages are triggered by an explicit request for 'status', 'bpm', etc?

As for the clock, that extra information was very helpful (and I'm surprised that it is not more obvious in the link documentation). It looks like time.monotonic() is going to work. I'll run a few experiments to see how accurate it is later today.